### PR TITLE
DRM Multi-GPU

### PR DESCRIPTION
**Pull request by ascent12**
Finally got this shit sorted.
This introduces the ability for multiple DRM backends to be running at the same time, and have each of them driving their own outputs.

This performs all of the rendering on the primary GPU ("boot_vga" or whichever was found first), and copies it over to the other GPU to be displayed.
It uses zero-copy (with DMA-BUFs and EGLImages), so it's not too inefficient. I didn't notice any performance problems in my testing.

We'll eventually need to have a way to specify which GPU should be doing the rendering, and maybe even have a way at runtime to switch between them (e.g. unplugging laptop will use iGPU instead of discrete GPU) if it's even possible. That can come later, though.

There is also various bits of refactoring in this PR.

### CODE REVIEW

File: backend/meson.build [hunk 1]
```diff
@@ -5,10 +5,10 @@ backend_files = files(
 	'session/session.c',
 	'drm/backend.c',
 	'drm/drm.c',
-	'drm/drm-atomic.c',
-	'drm/drm-legacy.c',
-	'drm/drm-properties.c',
-	'drm/drm-util.c',
+	'drm/iface_atomic.c',
+	'drm/iface_legacy.c',
```
Comment (SirCmpwn) [inline backend/meson.build]: I would just call these atomic.c and legacy.c

File: backend/drm/drm.c [hunk 2]
```diff
@@ -20,12 +20,12 @@
...
-#include "backend/drm.h"
-#include "backend/drm-util.h"
+#include "backend/drm/drm.h"
+#include "backend/drm/util.h"
 
 bool wlr_drm_check_features(struct wlr_drm_backend *backend) {
-	extern const struct wlr_drm_interface legacy_iface;
-	extern const struct wlr_drm_interface atomic_iface;
+	extern const struct wlr_drm_interface iface_legacy;
+	extern const struct wlr_drm_interface iface_atomic;
```
Comment (SirCmpwn) [inline backend/drm/drm.c]: And the old names for these variables made more sense imo.

File: backend/backend.c [hunk 3]
```diff
@@ -91,37 +91,46 @@ struct wlr_backend *wlr_backend_autocreate(struct wl_display *display) {
...
+		if (!primary_drm) {
+			primary_drm = drm;
+		}
+
+		wlr_multi_backend_add(backend, drm);
+	}
+
+	if (!primary_drm) {
+		wlr_log(L_ERROR, "Failed to open any DRM device");
+		goto error_multi;
```
Comment (SirCmpwn) [inline backend/backend.c]: Something that might be useful in the future is to specify a list of GPUs to use - then maybe you can run a different compositor on each GPU, for example.

**Comment by SirCmpwn** [comment]
Overall, this looks good. Excellent work. Thanks!

File: backend/meson.build (hunk 1 above)
Comment (ascent12) [reply backend/meson.build]: I wanted it to be obvious that they were related to iface.h.

File: backend/backend.c (hunk 3 above)
Comment (ascent12) [reply backend/backend.c]: Yes, being able to explicitly pick the GPUs would be helpful.
Maybe a colon-separated environment variable would be helpful, where the first one specified is the primary GPU:
WLR_DRM_DEVICES=/dev/dri/card0:/dev/dri/card3

File: backend/meson.build (hunk 1 above)
Comment (SirCmpwn) [reply backend/meson.build]: I think that was already obvious.

Comment (SirCmpwn) [reply backend/meson.build]: iface_atomic reads like hungarian notation imo

**Comment by SirCmpwn** [comment]
Excellent, thanks!

**Pull request merged**